File: plugin/src/write-styles.ts
import { makeSolidPaint } from "./write-helpers";
import { parseColor } from "./color";
import { assertNotFigjam } from "./figjam";

// Each style type accepts only its own fields — a fontSize on a paint style is a
// caller mistake, not a silent no-op.
const STYLE_FIELDS: Record<string, readonly string[]> = {
  PAINT: ["name", "description", "color"],
  TEXT: [
    "name", "description", "fontFamily", "fontStyle", "fontSize", "textDecoration",
    "lineHeightValue", "lineHeightUnit", "letterSpacingValue", "letterSpacingUnit",
  ],
  EFFECT: ["name", "description", "effects"],
  GRID: [
    "name", "description", "pattern", "count", "gutterSize", "offset", "alignment",
    "sectionSize", "color", "opacity",
  ],
};

const STYLE_BINDABLE_FIELDS: Record<string, readonly string[]> = {
  PAINT: ["color"],
  TEXT: [
    "fontFamily", "fontSize", "fontStyle", "fontWeight", "letterSpacing", "lineHeight",
    "paragraphSpacing", "paragraphIndent",
  ],
  EFFECT: ["color", "radius", "spread", "offsetX", "offsetY"],
  GRID: ["sectionSize", "count", "offset", "gutterSize"],
};

// Shared by set_effects and update_style: one array-entry effect schema.
const buildEffect = (e: any): Effect => {
  switch (e.type) {
    case "DROP_SHADOW":
    case "INNER_SHADOW": {
      const { r, g, b } = parseColor(e.color || "#000000");
      return {
        type: e.type as "DROP_SHADOW" | "INNER_SHADOW",
        color: { r, g, b, a: e.opacity != null ? Number(e.opacity) : 0.25 },
        offset: { x: Number(e.offsetX ?? 0), y: Number(e.offsetY ?? 4) },
        radius: Number(e.radius ?? 4),
        spread: Number(e.spread ?? 0),
        visible: e.visible ?? true,
        blendMode: (e.blendMode || "NORMAL") as BlendMode,
      } as DropShadowEffect;
    }
    case "LAYER_BLUR":
    case "BACKGROUND_BLUR":
      return {
        type: e.type as "LAYER_BLUR" | "BACKGROUND_BLUR",
        radius: Number(e.radius ?? 4),
        visible: e.visible ?? true,
      } as BlurEffect;
    default:
      throw new Error(`Unknown effect type: ${e.type}. Must be DROP_SHADOW, INNER_SHADOW, LAYER_BLUR, or BACKGROUND_BLUR`);
  }
};

// Shared by create_grid_style and update_style.
const buildLayoutGrid = (p: any): LayoutGrid => {
  const pattern = p.pattern || "GRID";
  if (pattern === "COLUMNS" || pattern === "ROWS") {
    return {
      pattern,
      count: Number(p.count ?? 12),
      gutterSize: Number(p.gutterSize ?? 16),
      offset: Number(p.offset ?? 0),
      alignment: p.alignment || "STRETCH",
      visible: true,
    };
  }
  const { r, g, b, a } = parseColor(p.color || "#FF0000");
  return {
    pattern: "GRID",
    sectionSize: Number(p.sectionSize ?? 8),
    visible: true,
    color: { r, g, b, a: p.opacity != null ? Number(p.opacity) : (a !== 1 ? a : 0.1) },
  };
};

export const handleWriteStyleRequest = async (request: any) => {
  switch (request.type) {
    case "create_paint_style": {
      assertNotFigjam("create_paint_style");
      const p = request.params || {};
      if (!p.name) throw new Error("name is required");
      if (!p.color) throw new Error("color is required");
      const existing = (await figma.getLocalPaintStylesAsync()).find(s => s.name === p.name);
      if (existing) {
        return { type: request.type, requestId: request.requestId, data: { id: existing.id, name: existing.name } };
      }
      const style = figma.createPaintStyle();
      style.name = p.name;
      style.paints = [makeSolidPaint(p.color)];
      if (p.description) style.description = p.description;
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: style.id, name: style.name },
      };
    }

    case "create_text_style": {
      assertNotFigjam("create_text_style");
      const p = request.params || {};
      if (!p.name) throw new Error("name is required");
      const existing = (await figma.getLocalTextStylesAsync()).find(s => s.name === p.name);
      if (existing) {
        return { type: request.type, requestId: request.requestId, data: { id: existing.id, name: existing.name } };
      }
      const family = p.fontFamily || "Inter";
      const fontStyle = p.fontStyle || "Regular";
      await figma.loadFontAsync({ family, style: fontStyle });
      const style = figma.createTextStyle();
      style.name = p.name;
      style.fontName = { family, style: fontStyle };
      if (p.fontSize != null) style.fontSize = Number(p.fontSize);
      if (p.description) style.description = p.description;
      if (p.textDecoration && p.textDecoration !== "NONE") {
        style.textDecoration = p.textDecoration;
      }
      if (p.lineHeightValue != null) {
        style.lineHeight = { value: Number(p.lineHeightValue), unit: p.lineHeightUnit || "PIXELS" };
      }
      if (p.letterSpacingValue != null) {
        style.letterSpacing = { value: Number(p.letterSpacingValue), unit: p.letterSpacingUnit || "PIXELS" };
      }
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: style.id, name: style.name },
      };
    }

    case "create_effect_style": {
      assertNotFigjam("create_effect_style");
      const p = request.params || {};
      if (!p.name) throw new Error("name is required");
      const existing = (await figma.getLocalEffectStylesAsync()).find(s => s.name === p.name);
      if (existing) {
        return { type: request.type, requestId: request.requestId, data: { id: existing.id, name: existing.name } };
      }
      const effectType = p.type || "DROP_SHADOW";
      let effect: Effect;
      if (effectType === "LAYER_BLUR") {
        effect = { type: "LAYER_BLUR", blurType: "NORMAL", radius: Number(p.radius ?? 4), visible: true };
      } else if (effectType === "BACKGROUND_BLUR") {
        effect = { type: "BACKGROUND_BLUR", blurType: "NORMAL", radius: Number(p.radius ?? 4), visible: true };
      } else {
        // DROP_SHADOW or INNER_SHADOW
        const { r, g, b, a } = parseColor(p.color || "#000000");
        const alpha = p.opacity != null ? Number(p.opacity) : (a !== 1 ? a : 0.25);
        effect = {
          type: effectType as "DROP_SHADOW" | "INNER_SHADOW",
          color: { r, g, b, a: alpha },
          offset: { x: Number(p.offsetX ?? 0), y: Number(p.offsetY ?? 4) },
          radius: Number(p.radius ?? 8),
          spread: Number(p.spread ?? 0),
          visible: true,
          blendMode: "NORMAL",
        };
      }
      const style = figma.createEffectStyle();
      style.name = p.name;
      style.effects = [effect];
      if (p.description) style.description = p.description;
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: style.id, name: style.name },
      };
    }

    case "create_grid_style": {
      const p = request.params || {};
      assertNotFigjam("create_grid_style");
      if (!p.name) throw new Error("name is required");
      const existing = (await figma.getLocalGridStylesAsync()).find(s => s.name === p.name);
      if (existing) {
        return { type: request.type, requestId: request.requestId, data: { id: existing.id, name: existing.name } };
      }
      const grid = buildLayoutGrid(p);
      const style = figma.createGridStyle();
      style.name = p.name;
      style.layoutGrids = [grid];
      if (p.description) style.description = p.description;
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: style.id, name: style.name },
      };
    }

    case "update_style": {
      const p = request.params || {};
      assertNotFigjam("update_style");
      if (!p.styleId) throw new Error("styleId is required");
      const style = await figma.getStyleByIdAsync(p.styleId);
      if (!style) throw new Error(`Style not found: ${p.styleId}`);
      const allowed = STYLE_FIELDS[style.type];
      if (!allowed) throw new Error(`Unknown style type: ${style.type}`);
      for (const key of Object.keys(p)) {
        if (key !== "styleId" && !allowed.includes(key)) {
          throw new Error(`${key} is not applicable to a ${style.type} style`);
        }
      }
      if (p.name != null) style.name = p.name;
      if (p.description != null) style.description = p.description;
      if (style.type === "PAINT") {
        if (p.color != null) (style as PaintStyle).paints = [makeSolidPaint(p.color)];
      } else if (style.type === "TEXT") {
        const text = style as TextStyle;
        if (p.fontFamily != null || p.fontStyle != null) {
          const family = p.fontFamily != null ? p.fontFamily : text.fontName.family;
          const fontStyle = p.fontStyle != null ? p.fontStyle : text.fontName.style;
          await figma.loadFontAsync({ family, style: fontStyle });
          text.fontName = { family, style: fontStyle };
        }
        if (p.fontSize != null) text.fontSize = Number(p.fontSize);
        if (p.textDecoration != null) text.textDecoration = p.textDecoration;
        if (p.lineHeightValue != null) {
          text.lineHeight = { value: Number(p.lineHeightValue), unit: p.lineHeightUnit || "PIXELS" };
        }
        if (p.letterSpacingValue != null) {
          text.letterSpacing = { value: Number(p.letterSpacingValue), unit: p.letterSpacingUnit || "PIXELS" };
        }
      } else if (style.type === "EFFECT") {
        if (p.effects != null) {
          if (!Array.isArray(p.effects)) throw new Error("effects array is required");
          (style as EffectStyle).effects = p.effects.map(buildEffect);
        }
      } else {
        const gridStyle = style as GridStyle;
        const current: any = gridStyle.layoutGrids[0];
        // A new pattern means a differently shaped grid — rebuild it. Otherwise
        // patch the existing one so untouched fields survive.
        if (p.pattern != null || !current) {
          gridStyle.layoutGrids = [buildLayoutGrid(p)];
        } else {
          const next: any = { ...current };
          if (p.count != null) next.count = Number(p.count);
          if (p.gutterSize != null) next.gutterSize = Number(p.gutterSize);
          if (p.offset != null) next.offset = Number(p.offset);
          if (p.alignment != null) next.alignment = p.alignment;
          if (p.sectionSize != null) next.sectionSize = Number(p.sectionSize);
          if (p.color != null || p.opacity != null) {
            const base = p.color != null
              ? parseColor(p.color)
              : { ...(current.color || { r: 1, g: 0, b: 0, a: 0.1 }) };
            next.color = {
              r: base.r,
              g: base.g,
              b: base.b,
              a: p.opacity != null ? Number(p.opacity) : base.a,
            };
          }
          gridStyle.layoutGrids = [next];
        }
      }
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: style.id, name: style.name, type: style.type },
      };
    }

    case "bind_variable_to_style": {
      const p = request.params || {};
      assertNotFigjam("bind_variable_to_style");
      if (!p.styleId) throw new Error("styleId is required");
      if (!p.field) throw new Error("field is required");
      const style = await figma.getStyleByIdAsync(p.styleId);
      if (!style) throw new Error(`Style not found: ${p.styleId}`);
      const variable = p.variableId
        ? await figma.variables.getVariableByIdAsync(p.variableId)
        : null;
      if (p.variableId && !variable) throw new Error(`Variable not found: ${p.variableId}`);
      const bindable = STYLE_BINDABLE_FIELDS[style.type];
      if (!bindable) throw new Error(`Unknown style type: ${style.type}`);
      if (!bindable.includes(p.field)) {
        throw new Error(
          `field ${p.field} is not bindable on a ${style.type} style — expected ${bindable.join(", ")}`,
        );
      }
      if (style.type === "PAINT") {
        const paintStyle = style as PaintStyle;
        const paints = [...paintStyle.paints];
        const base = paints.length > 0 ? paints[0] : makeSolidPaint("#000000");
        if (base.type !== "SOLID") {
          throw new Error(`Style ${p.styleId} paint 0 is ${base.type}, not SOLID`);
        }
        paints[0] = figma.variables.setBoundVariableForPaint(base, "color", variable);
        paintStyle.paints = paints;
      } else if (style.type === "TEXT") {
        (style as TextStyle).setBoundVariable(p.field as VariableBindableTextField, variable);
      } else if (style.type === "EFFECT") {
        const effectStyle = style as EffectStyle;
        const effects = [...effectStyle.effects];
        if (effects.length === 0) throw new Error(`Style ${p.styleId} has no effects to bind`);
        effects[0] = figma.variables.setBoundVariableForEffect(
          effects[0],
          p.field as VariableBindableEffectField,
          variable,
        );
        effectStyle.effects = effects;
      } else {
        const gridStyle = style as GridStyle;
        const grids = [...gridStyle.layoutGrids];
        if (grids.length === 0) throw new Error(`Style ${p.styleId} has no layout grids to bind`);
        grids[0] = figma.variables.setBoundVariableForLayoutGrid(
          grids[0],
          p.field as VariableBindableLayoutGridField,
          variable,
        );
        gridStyle.layoutGrids = grids;
      }
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: {
          styleId: style.id,
          styleType: style.type,
          field: p.field,
          variableId: p.variableId != null ? p.variableId : null,
          bound: variable !== null,
        },
      };
    }

    case "delete_style": {
      const p = request.params || {};
      assertNotFigjam("delete_style");
      if (!p.styleId) throw new Error("styleId is required");
      const style = await figma.getStyleByIdAsync(p.styleId);
      if (!style) throw new Error(`Style not found: ${p.styleId}`);
      style.remove();
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { styleId: p.styleId, deleted: true },
      };
    }

    case "apply_style_to_node": {
      const p = request.params || {};
      assertNotFigjam("apply_style_to_node");
      const nodeId = request.nodeIds && request.nodeIds[0];
      if (!nodeId) throw new Error("nodeId is required");
      if (!p.styleId) throw new Error("styleId is required");
      const node = await figma.getNodeByIdAsync(nodeId);
      if (!node) throw new Error(`Node not found: ${nodeId}`);
      const style = await figma.getStyleByIdAsync(p.styleId);
      if (!style) throw new Error(`Style not found: ${p.styleId}`);
      const n = node as any;
      switch (style.type) {
        case "PAINT": {
          const target = p.target || "fill";
          if (target === "stroke") {
            if (!("strokeStyleId" in node)) throw new Error(`Node ${nodeId} does not support stroke styles`);
            await n.setStrokeStyleIdAsync(p.styleId);
          } else {
            if (!("fillStyleId" in node)) throw new Error(`Node ${nodeId} does not support fill styles`);
            await n.setFillStyleIdAsync(p.styleId);
          }
          break;
        }
        case "TEXT":
          if (!("textStyleId" in node)) throw new Error(`Node ${nodeId} does not support text styles`);
          await n.setTextStyleIdAsync(p.styleId);
          break;
        case "EFFECT":
          if (!("effectStyleId" in node)) throw new Error(`Node ${nodeId} does not support effect styles`);
          await n.setEffectStyleIdAsync(p.styleId);
          break;
        case "GRID":
          if (!("gridStyleId" in node)) throw new Error(`Node ${nodeId} does not support grid styles`);
          await n.setGridStyleIdAsync(p.styleId);
          break;
        default:
          throw new Error(`Unknown style type: ${(style as any).type}`);
      }
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: n.id, name: n.name, styleId: p.styleId, styleType: style.type },
      };
    }

    case "set_effects": {
      const p = request.params || {};
      assertNotFigjam("set_effects");
      const nodeId = request.nodeIds && request.nodeIds[0];
      if (!nodeId) throw new Error("nodeId is required");
      if (!Array.isArray(p.effects)) throw new Error("effects array is required");
      const node = await figma.getNodeByIdAsync(nodeId) as any;
      if (!node) throw new Error(`Node not found: ${nodeId}`);
      if (!("effects" in node)) throw new Error(`Node ${nodeId} does not support effects`);
      const effects: Effect[] = p.effects.map(buildEffect);
      node.effects = effects;
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: { id: node.id, name: node.name, effectCount: effects.length },
      };
    }

    case "bind_variable_to_node": {
      const p = request.params || {};
      assertNotFigjam("bind_variable_to_node");
      const nodeId = request.nodeIds && request.nodeIds[0];
      if (!nodeId) throw new Error("nodeId is required");
      if (!p.field) throw new Error("field is required");
      const node = await figma.getNodeByIdAsync(nodeId) as any;
      if (!node) throw new Error(`Node not found: ${nodeId}`);
      const variable = p.variableId
        ? await figma.variables.getVariableByIdAsync(p.variableId)
        : null;
      if (p.variableId && !variable) throw new Error(`Variable not found: ${p.variableId}`);
      if (p.field === "fillColor") {
        if (!("fills" in node)) throw new Error(`Node ${nodeId} does not support fills`);
        const fills = [...(node.fills as Paint[])];
        const base = fills.length > 0 ? fills[0] : makeSolidPaint("#000000");
        if (base.type !== "SOLID") throw new Error(`Node ${nodeId} fill 0 is ${base.type}, not SOLID`);
        fills[0] = figma.variables.setBoundVariableForPaint(base as SolidPaint, "color", variable);
        node.fills = fills;
      } else if (p.field === "strokeColor") {
        if (!("strokes" in node)) throw new Error(`Node ${nodeId} does not support strokes`);
        const strokes = [...(node.strokes as Paint[])];
        const base = strokes.length > 0 ? strokes[0] : makeSolidPaint("#000000");
        if (base.type !== "SOLID") throw new Error(`Node ${nodeId} stroke 0 is ${base.type}, not SOLID`);
        strokes[0] = figma.variables.setBoundVariableForPaint(base as SolidPaint, "color", variable);
        node.strokes = strokes;
      } else {
        if (!(p.field in node)) throw new Error(`Node ${nodeId} does not have field: ${p.field}`);
        node.setBoundVariable(p.field, variable);
      }
      figma.commitUndo();
      return {
        type: request.type,
        requestId: request.requestId,
        data: {
          id: node.id,
          name: node.name,
          variableId: p.variableId != null ? p.variableId : null,
          field: p.field,
          bound: variable !== null,
        },
      };
    }

    default:
      return null;
  }
};

File: src/tools/write_styles.rs
// MANIFEST: create_paint_style | CreatePaintStyleArgs | Create a new local paint style with a solid fill color (hex, rgb(), hsl(), or oklch()).
// MANIFEST: create_text_style | CreateTextStyleArgs | Create a new local text style (typography preset). Returns the new style's ID. Apply it to nodes with apply_style_to_node. Use get_styles to list existing text styles.
// MANIFEST: create_effect_style | CreateEffectStyleArgs | Create a new local effect style (drop shadow, inner shadow, or blur).
// MANIFEST: create_grid_style | CreateGridStyleArgs | Create a new local layout grid style.
// MANIFEST: update_style | UpdateStyleArgs | Update an existing style in place: paint (color), text (font, size, line height, letter spacing, decoration), effect (effects array), or grid (pattern, count, gutter, section size). Also renames and re-describes any style type. Pass only fields that apply to that style's type.
// MANIFEST: delete_style | DeleteStyleArgs | Delete a style (paint, text, effect, or grid) by its ID.
// MANIFEST: apply_style_to_node | ApplyStyleToNodeArgs | Apply an existing local style (paint, text, effect, or grid) to a node, linking the node to that style.
// MANIFEST: set_effects | SetEffectsArgs | Apply one or more effects (drop shadow, inner shadow, layer blur, background blur) directly to a node. Replaces all existing effects. Pass an empty array to clear all effects.
// MANIFEST: bind_variable_to_node | BindVariableToNodeArgs | Bind a local variable to a node property so the property is driven by the variable's value. COLOR variables: use fillColor or strokeColor. BOOLEAN variables: use visible. FLOAT variables: use opacity, width, height, cornerRadius, strokeWeight, itemSpacing, paddingTop, paddingRight, paddingBottom, paddingLeft, and the other numeric fields listed on the field parameter. Omit variableId to unbind the field. Binding fillColor/strokeColor rewrites only the first paint and leaves the rest of the array intact. Note — binding cornerRadius fans out to all four corner fields, so unbinding it requires one call per corner (topLeftRadius, topRightRadius, bottomLeftRadius, bottomRightRadius).
// MANIFEST: bind_variable_to_style | BindVariableToStyleArgs | Bind a variable to a local style so the style itself is driven by a token: paint styles accept field 'color'; text styles accept fontFamily, fontSize, fontStyle, fontWeight, lineHeight, letterSpacing, paragraphSpacing, paragraphIndent; effect styles accept color, radius, spread, offsetX, offsetY; grid styles accept sectionSize, count, offset, gutterSize. Omit variableId to unbind.

//! Style tools: paint/text/effect/grid style creation, updates, and application.

use std::sync::Arc;

use super::McpError;
use rmcp::model::CallToolResult;
use rmcp::schemars;
use serde::{Deserialize, Serialize};

use crate::node::Node;

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreatePaintStyleArgs {
    /// Style name e.g. 'Brand/Primary'
    pub name: String,
    /// Fill color as hex e.g. #FF5733
    pub color: String,
    /// Optional style description
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateTextStyleArgs {
    /// Style name — use slash notation to organise into groups e.g. 'Heading/H1', 'Body/Regular'
    pub name: String,
    /// Font size in pixels (default 16)
    #[serde(default)]
    pub font_size: Option<f64>,
    /// Font family name e.g. 'Inter', 'Roboto' (default Inter). Must be installed in Figma.
    #[serde(default)]
    pub font_family: Option<String>,
    /// Font style variant e.g. 'Regular', 'Bold', 'Medium', 'SemiBold' (default Regular)
    #[serde(default)]
    pub font_style: Option<String>,
    /// Text decoration: NONE (default), UNDERLINE, or STRIKETHROUGH
    #[serde(default)]
    pub text_decoration: Option<String>,
    /// Line height value (unit set by lineHeightUnit)
    #[serde(default)]
    pub line_height_value: Option<f64>,
    /// Line height unit: PIXELS (default) or PERCENT
    #[serde(default)]
    pub line_height_unit: Option<String>,
    /// Letter spacing value (unit set by letterSpacingUnit)
    #[serde(default)]
    pub letter_spacing_value: Option<f64>,
    /// Letter spacing unit: PIXELS (default) or PERCENT
    #[serde(default)]
    pub letter_spacing_unit: Option<String>,
    /// Optional human-readable description shown in the Figma style panel
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateEffectStyleArgs {
    /// Style name e.g. 'Shadow/Card'
    pub name: String,
    /// Effect type: DROP_SHADOW (default), INNER_SHADOW, LAYER_BLUR, or BACKGROUND_BLUR
    #[serde(default, rename = "type")]
    pub r#type: Option<String>,
    /// Shadow color as hex e.g. #000000 (default #000000, shadows only)
    #[serde(default)]
    pub color: Option<String>,
    /// Shadow color opacity 0–1 (default 0.25, shadows only)
    #[serde(default)]
    pub opacity: Option<f64>,
    /// Blur radius in pixels (default 8 for shadows, 4 for blurs)
    #[serde(default)]
    pub radius: Option<f64>,
    /// Shadow X offset in pixels (default 0, shadows only)
    #[serde(default)]
    pub offset_x: Option<f64>,
    /// Shadow Y offset in pixels (default 4, shadows only)
    #[serde(default)]
    pub offset_y: Option<f64>,
    /// Shadow spread in pixels (default 0, shadows only)
    #[serde(default)]
    pub spread: Option<f64>,
    /// Optional style description
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct CreateGridStyleArgs {
    /// Style name e.g. 'Grid/Desktop'
    pub name: String,
    /// Grid pattern: GRID (default), COLUMNS, or ROWS
    #[serde(default)]
    pub pattern: Option<String>,
    /// Number of columns or rows (COLUMNS/ROWS only, default 12)
    #[serde(default)]
    pub count: Option<f64>,
    /// Gutter size in pixels (COLUMNS/ROWS only, default 16)
    #[serde(default)]
    pub gutter_size: Option<f64>,
    /// Margin/offset in pixels (COLUMNS/ROWS only, default 0)
    #[serde(default)]
    pub offset: Option<f64>,
    /// Alignment: STRETCH (default), CENTER, MIN, or MAX (COLUMNS/ROWS only)
    #[serde(default)]
    pub alignment: Option<String>,
    /// Grid cell size in pixels (GRID only, default 8)
    #[serde(default)]
    pub section_size: Option<f64>,
    /// Grid line color as hex, rgb(), hsl(), or oklch() e.g. #FF0000, hsl(0 100% 50%), oklch(0.63 0.26 29) (GRID only, default #FF0000)
    #[serde(default)]
    pub color: Option<String>,
    /// Grid line opacity 0–1 (GRID only, default 0.1)
    #[serde(default)]
    pub opacity: Option<f64>,
    /// Optional style description
    #[serde(default)]
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct UpdateStyleArgs {
    /// Style ID to update (from get_styles)
    pub style_id: String,
    /// New style name (any style type)
    #[serde(default)]
    pub name: Option<String>,
    /// New style description (any style type)
    #[serde(default)]
    pub description: Option<String>,
    /// New color as hex, rgb(), hsl(), or oklch() (PAINT: the fill; GRID: the grid line color)
    #[serde(default)]
    pub color: Option<String>,
    /// New font family (TEXT only) — loaded before it is applied
    #[serde(default)]
    pub font_family: Option<String>,
    /// New font style e.g. 'Bold' (TEXT only)
    #[serde(default)]
    pub font_style: Option<String>,
    /// New font size in pixels (TEXT only)
    #[serde(default)]
    pub font_size: Option<f64>,
    /// Text decoration: NONE, UNDERLINE, or STRIKETHROUGH (TEXT only)
    #[serde(default)]
    pub text_decoration: Option<String>,
    /// Line height value (TEXT only)
    #[serde(default)]
    pub line_height_value: Option<f64>,
    /// Line height unit: PIXELS (default) or PERCENT (TEXT only)
    #[serde(default)]
    pub line_height_unit: Option<String>,
    /// Letter spacing value (TEXT only)
    #[serde(default)]
    pub letter_spacing_value: Option<f64>,
    /// Letter spacing unit: PIXELS (default) or PERCENT (TEXT only)
    #[serde(default)]
    pub letter_spacing_unit: Option<String>,
    /// Replacement effects array, same schema as set_effects (EFFECT only)
    #[serde(default)]
    pub effects: Option<Vec<serde_json::Value>>,
    /// Grid pattern: GRID, COLUMNS, or ROWS (GRID only) — supplying it rebuilds the grid
    #[serde(default)]
    pub pattern: Option<String>,
    /// Number of columns or rows (GRID only, COLUMNS/ROWS patterns)
    #[serde(default)]
    pub count: Option<f64>,
    /// Gutter size in pixels (GRID only, COLUMNS/ROWS patterns)
    #[serde(default)]
    pub gutter_size: Option<f64>,
    /// Margin/offset in pixels (GRID only, COLUMNS/ROWS patterns)
    #[serde(default)]
    pub offset: Option<f64>,
    /// Alignment: STRETCH, CENTER, MIN, or MAX (GRID only, COLUMNS/ROWS patterns)
    #[serde(default)]
    pub alignment: Option<String>,
    /// Grid cell size in pixels (GRID only, GRID pattern)
    #[serde(default)]
    pub section_size: Option<f64>,
    /// Grid line opacity 0–1 (GRID only, GRID pattern)
    #[serde(default)]
    pub opacity: Option<f64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct DeleteStyleArgs {
    /// Style ID to delete
    pub style_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct ApplyStyleToNodeArgs {
    /// Target node ID in colon format e.g. 4029:12345
    pub node_id: String,
    /// Style ID to apply (from get_styles)
    pub style_id: String,
    /// For paint styles only — apply to 'fill' (default) or 'stroke'
    #[serde(default)]
    pub target: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct SetEffectsArgs {
    /// Target node ID in colon format e.g. 4029:12345
    pub node_id: String,
    /// Array of effect objects. Each has: type (DROP_SHADOW | INNER_SHADOW | LAYER_BLUR | BACKGROUND_BLUR), radius, color (hex, rgb(), hsl(), or oklch(); shadows only), opacity (0–1, shadows only), offsetX, offsetY (shadows only), spread (shadows only), visible (default true)
    pub effects: Vec<serde_json::Value>,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct BindVariableToNodeArgs {
    /// Target node ID in colon format e.g. 4029:12345
    pub node_id: String,
    /// Variable ID to bind (from get_variable_defs). Omit to unbind the field and leave its current value in place.
    #[serde(default)]
    pub variable_id: Option<String>,
    /// Property to bind: fillColor | strokeColor | visible | opacity | width | height | minWidth | maxWidth | minHeight | maxHeight | cornerRadius | topLeftRadius | topRightRadius | bottomLeftRadius | bottomRightRadius | strokeWeight | strokeTopWeight | strokeRightWeight | strokeBottomWeight | strokeLeftWeight | itemSpacing | counterAxisSpacing | paddingTop | paddingRight | paddingBottom | paddingLeft | gridRowGap | gridColumnGap | characters | fontFamily | fontSize | fontStyle | fontWeight | lineHeight | letterSpacing | paragraphSpacing | paragraphIndent
    pub field: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, schemars::JsonSchema)]
#[serde(rename_all = "camelCase")]
pub(crate) struct BindVariableToStyleArgs {
    /// Style ID to bind (from get_styles)
    pub style_id: String,
    /// Field to bind. PAINT: color. TEXT: fontFamily | fontSize | fontStyle | fontWeight | lineHeight | letterSpacing | paragraphSpacing | paragraphIndent. EFFECT: color | radius | spread | offsetX | offsetY. GRID: sectionSize | count | offset | gutterSize.
    pub field: String,
    /// Variable ID to bind (from get_variable_defs). Omit to unbind the field.
    #[serde(default)]
    pub variable_id: Option<String>,
}

pub(crate) async fn create_paint_style(
    node: Arc<Node>,
    args: CreatePaintStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "create_paint_style", &args).await
}

pub(crate) async fn create_text_style(
    node: Arc<Node>,
    args: CreateTextStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "create_text_style", &args).await
}

pub(crate) async fn create_effect_style(
    node: Arc<Node>,
    args: CreateEffectStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "create_effect_style", &args).await
}

pub(crate) async fn create_grid_style(
    node: Arc<Node>,
    args: CreateGridStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "create_grid_style", &args).await
}

pub(crate) async fn update_style(
    node: Arc<Node>,
    args: UpdateStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "update_style", &args).await
}

pub(crate) async fn delete_style(
    node: Arc<Node>,
    args: DeleteStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "delete_style", &args).await
}

pub(crate) async fn apply_style_to_node(
    node: Arc<Node>,
    args: ApplyStyleToNodeArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "apply_style_to_node", &args).await
}

pub(crate) async fn set_effects(
    node: Arc<Node>,
    args: SetEffectsArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "set_effects", &args).await
}

pub(crate) async fn bind_variable_to_node(
    node: Arc<Node>,
    args: BindVariableToNodeArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "bind_variable_to_node", &args).await
}

pub(crate) async fn bind_variable_to_style(
    node: Arc<Node>,
    args: BindVariableToStyleArgs,
) -> Result<CallToolResult, McpError> {
    crate::tools::relay(&node, "bind_variable_to_style", &args).await
}
